### ISSUE TITLE
Private array behavior depends on backend

### ISSUE THREAD

**Issue by mwarusz** [JuliaGPU/KernelAbstractions.jl#142]
Private arrays behave differently depending on the chosen backend when accessed from functions inside a kernel. For example
```
using KernelAbstractions, CUDA

function setpriv!(a, val)
  @inbounds a[1] = val
end

@kernel function testkernel!(A)
  I = @index(Global)
  priv = @private Int 1
  setpriv!(priv, 1)
  @inbounds A[I] = priv[1]
end

let
  A = zeros(Int, 8)
  wait(testkernel!(CPU(), 4, 8)(A))
  @show A
end

let
  A = CuArray(zeros(Int, 8))
  wait(testkernel!(CUDADevice(), 4, 8)(A))
  @show Array(A)
end
```
gives
```
A = [1, 0, 0, 0, 1, 0, 0, 0]
Array(A) = [1, 1, 1, 1, 1, 1, 1, 1]
```
This is because on the CPU the `@kernel` macro adds extra indices only to private array accesses in the kernel main body. My (flawed) attempt to fix that was in #98, but I forgot to create an issue to track this.

**Comment by vchuravy** [comment]
Yeah... part of the issue we should in all likelihood rewrite the arrays to be passed as views. Or we need to rewrite the indexing operations with Cassette instead.

**Comment by vchuravy** [comment]
Should be fixed by #175 now private memory is implemented as a ntuple, and every use is down by indexing.
